### PR TITLE
Auton recorder

### PR DESCRIPTION
**Pull request by kirstenalice**
beginnings of the auton recorder
not finished, string to frame and frame to string conversion is functional
reading and writing to file is not done and not tested

### CODE REVIEW

File: 2016-2017_Starstruck/Module Test Code/FileIO_Test/include/AutonRecorder.h [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+ *   a file in the cortex's memory, and is able to read the file again to replay it as an
+ *   autonomous program. It can work for any game.
+ *
+ * This file should be placed in include and included in main.h
+ */
+
+
+  typedef struct Frame {
+      int analog_main[4]; //ch1, ch2, ch3, ch4
+      bool digital_main[12]; //5U, 5D, 6U, 6D, 7U, 7D, 7L, 7R, 8U, 8D, 8L, 8R
```
Comment (vex750evolution) [inline 2016-2017_Starstruck/Module Test Code/FileIO_Test/include/AutonRecorder.h]: Might want to add macros for these to make it easier so we can do something like frame.digital_main[J_5U] to get values.

File: 2016-2017_Starstruck/Module Test Code/FileIO_Test/include/AutonRecorder.h [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+      int analog_partner[4];
+      bool digital_partner[12];
+      struct Frame *next;
+      struct Frame *previous;
+    }Frame;
+
+
+  static const Frame NULLFRAME = {
+    {-1,-1,-1,-1},
```
Comment (vex750evolution) [inline 2016-2017_Starstruck/Module Test Code/FileIO_Test/include/AutonRecorder.h]: when defining structs, you want to try to use the more explicit syntax when possible. i.e.:

```
static const Frame NULLFRAME = {
    .analog_main = {-1, -1, -1, -1},
...
```

That way you can easily tell what's being assigned to what and you don't have to memorize the struct order

File: 2016-2017_Starstruck/Module Test Code/FileIO_Test/src/AutonRecorder.c [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+  void addFrame(Frame toAdd){
+      toAdd.previous = root.previous;
+      toAdd.next = &root;
+      root.previous = &toAdd;
+  }
+
+  char * frameToString(Frame frame){
+    char * string = malloc(50);
+    sprintf(string,"%d%d%d%d%d%d%d%d%d%d%d%d%d%d%d%d%d%d%d%d%d%d%d%d%d%d%d%d%d%d%d%d",
+      frame.analog_main[0]+127,
```
Comment (vex750evolution) [inline 2016-2017_Starstruck/Module Test Code/FileIO_Test/src/AutonRecorder.c]: see comment in header file about macros, these array indeces are not self-explanatory at all

File: 2016-2017_Starstruck/Module Test Code/FileIO_Test/src/AutonRecorder.c [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+  }
+
+  void addFrame(Frame toAdd){
+      toAdd.previous = root.previous;
+      toAdd.next = &root;
+      root.previous = &toAdd;
+  }
+
+  char * frameToString(Frame frame){
+    char * string = malloc(50);
```
Comment (vex750evolution) [inline 2016-2017_Starstruck/Module Test Code/FileIO_Test/src/AutonRecorder.c]: don't hardcode number of bytes. Use `malloc(sizeof(char)*strLength)`

File: 2016-2017_Starstruck/Module Test Code/FileIO_Test/src/AutonRecorder.c [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+    char* analog_main = substring(string,1,12);
+    char* digital_main = substring(string,13,12);
+    char* analog_partner = substring(string,25,12);
+    char* digital_partner = substring(string,37,12);
+
+    // printf("%d\t%d\t%d\t%d\t",frame.analog_main[0],frame.analog_main[1],frame.analog_main[2],frame.analog_main[3]);
+
+    int x;
+    for(x=0;x<4;x++){
+      frame.analog_main[x] = atoi(substring(analog_main,(x*3)+1,3))-127;
```
Comment (vex750evolution) [inline 2016-2017_Starstruck/Module Test Code/FileIO_Test/src/AutonRecorder.c]: might want to add comments to explain this math cuz we're not gonna remember what it means later

File: 2016-2017_Starstruck/Module Test Code/FileIO_Test/src/AutonRecorder.c [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+
+    char *substring = malloc(length+1);
+    int subPos;
+
+    if(substring == NULL){
+       printf("str in substring method is NULL.\n");
+       exit(1);
+    }
+
+    for(subPos = 0 ; subPos < length ; subPos++){
```
Comment (vex750evolution) [inline 2016-2017_Starstruck/Module Test Code/FileIO_Test/src/AutonRecorder.c]: not a huge fan of this implementation, I think the strncpy version on #electrical will be cleaner and faster (worth it to benchmark speeds or nah?)

File: 2016-2017_Starstruck/Module Test Code/FileIO_Test/src/opcontrol.c [hunk 7]
```diff
@@ -38,8 +38,7 @@
  * ======== */
 
 void operatorControl() {
```
Comment (vex750evolution) [inline 2016-2017_Starstruck/Module Test Code/FileIO_Test/src/opcontrol.c]: I'm assuming all of this is just test code?

File: 2016-2017_Starstruck/Module Test Code/FileIO_Test/include/AutonRecorder.h [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+    {false,false,false,false,false,false,false,false,false,false,false,false},
+    {-1,-1,-1,-1},
+    {false,false,false,false,false,false,false,false,false,false,false,false},
+    (Frame*)0,(Frame*)0
+  };
+
+  void startRecording(int slot);
+  /*
```
Comment (vex750evolution) [inline 2016-2017_Starstruck/Module Test Code/FileIO_Test/include/AutonRecorder.h]: docs should go above function, and use @param annotations to explain parameters (see docs format in API.h or LCD.h)  The reason for this is that the IDE will pick up on it and show the explanations in the autocomplete popups. We can also use it later to auto-generate an API reference for the modules

File: 2016-2017_Starstruck/Module Test Code/FileIO_Test/include/main.h [hunk 9]
```diff
@@ -44,6 +44,9 @@
 #define MAIN_H_
 
 #include <API.h>
+#include <AutonRecorder.h>
+#include <stdlib.h>
```
Comment (vex750evolution) [inline 2016-2017_Starstruck/Module Test Code/FileIO_Test/include/main.h]: I think stdlib is included in API.h, but not sure. string.h should be included in AutonRecorder.h to make sure dependencies are managed properly (if this is being imported into a new project, you don't want the user to have to worry about adding system imports that your module is dependent on)

File: 2016-2017_Starstruck/Module Test Code/FileIO_Test/src/AutonRecorder.c [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+ *  AutonRecorder.c
+ *
+ *  Created on: Feb 8, 2017
+ *      Author: Kirsten Meidlinger
+ */
+
+ #include <main.h>
+
+/*
+ * This file abstracts all of the functionality of 750E's Auton Recorder.
```
Comment (vex750evolution) [inline 2016-2017_Starstruck/Module Test Code/FileIO_Test/src/AutonRecorder.c:11]: File descriptions are gonna have to be tweaked but that's a later problem

File: 2016-2017_Starstruck/Module Test Code/FileIO_Test/src/AutonRecorder.c [hunk 11]
```diff
@@ -0,0 +1,193 @@
...
+    };
+    frame.next=&frame;
+    frame.previous=&frame;
+    return frame;
+  }
+
+  void addFrame(Frame toAdd){
+      toAdd.previous = root.previous;
+      toAdd.next = &root;
+      root.previous = &toAdd;
```
Comment (vex750evolution) [inline 2016-2017_Starstruck/Module Test Code/FileIO_Test/src/AutonRecorder.c]: need to get root.previous before changing so you can set its next to toAdd, you're missing a link

File: 2016-2017_Starstruck/Module Test Code/FileIO_Test/src/AutonRecorder.c [hunk 12]
```diff
@@ -0,0 +1,193 @@
...
+    //Close file
+    fclose(f);
+    printf("%d, %d, %d\n\r",atoi(a),atoi(b),atoi(c));
+    */
+
+    return root; //TODO: placeholder
+  }
+
+  void recordingLoop(){
+    if(root.analog_main[0] == -1){ //checks if it's equal to NULLFRAME
```
Comment (vex750evolution) [inline 2016-2017_Starstruck/Module Test Code/FileIO_Test/src/AutonRecorder.c]: any reason you did this here and not in startRecording? Might actually be better this way but I was wondering if you had a specific reason

**Review by vex750evolution — COMMENTED** [review]
Looks great so far, just commented in a few notes and some convention/semantics fixes. 

Reviewed by @AsciiPorpoise 
(Forgot I was on the team account when I started writing this review and I don't want to redo it)

File: 2016-2017_Starstruck/Module Test Code/FileIO_Test/src/AutonRecorder.c (hunk 6 above)
Comment (kirstenalice) [reply 2016-2017_Starstruck/Module Test Code/FileIO_Test/src/AutonRecorder.c]: i tried using it but it doesnt work lol

Comment (matticoli) [reply 2016-2017_Starstruck/Module Test Code/FileIO_Test/src/AutonRecorder.c]: I threw the code I sent together using nano and gcc, didn't test on the cortex. If this works we can take a look at it later, I'll open an issue for it

Comment (matticoli) [reply 2016-2017_Starstruck/Module Test Code/FileIO_Test/src/AutonRecorder.c]: See #30

Comment (kirstenalice) [reply 2016-2017_Starstruck/Module Test Code/FileIO_Test/src/AutonRecorder.c]: this is what u gave me before it doesn't work we can look at it tomorrow

File: 2016-2017_Starstruck/Module Test Code/FileIO_Test/include/AutonRecorder.h (hunk 8 above)
Comment (kirstenalice) [reply 2016-2017_Starstruck/Module Test Code/FileIO_Test/include/AutonRecorder.h]: idk what ur talking about

File: 2016-2017_Starstruck/Module Test Code/FileIO_Test/src/AutonRecorder.c (hunk 12 above)
Comment (kirstenalice) [reply 2016-2017_Starstruck/Module Test Code/FileIO_Test/src/AutonRecorder.c]: did what? checking if the root is null or assigning the root a value?

Comment (matticoli) [reply 2016-2017_Starstruck/Module Test Code/FileIO_Test/src/AutonRecorder.c]: assigning it a value. This actually works better so you don't have to pass a reference to the root out of the init function (see discussion on #28 )

Comment (kirstenalice) [reply 2016-2017_Starstruck/Module Test Code/FileIO_Test/src/AutonRecorder.c]: idk what youre talking about

**Review by matticoli — APPROVED** [review]

**Pull request merged**